src/Formik.tsx

```diff
@@ -324,14 +324,17 @@ export class Formik<Values = {}, ExtraProps = {}> extends React.Component<
 
       if (field) {
         // Set form fields by name
-        this.setState(prevState => ({
-          ...prevState,
-          values: setIn(prevState.values, field!, val),
-        }));
-
-        if (this.props.validateOnChange) {
-          this.runValidations(setIn(this.state.values, field, val));
```

Comment:
As the above code sets the values, this code used to run with outdated values in certain cases. We may only run the validation once the state update has been commited, so we move the validation into the callback.